### ISSUE TITLE
rustdoc for `(a or b or c) and a` should cut unnecessary conditions

### ISSUE THREAD

**Issue by ventaquil** [rust-lang/rust#104991]
I tried this code:

```rust
#[cfg(
    doc(
        cfg(
            any(
                feature = "a",
                feature = "b",
                feature = "c",
            )
        )
    )
)]
#[cfg(
    doc(
        cfg(
            feature = "a",
        )
    )
)]
pub fn add(left: usize, right: usize) -> usize {
    left + right
}
```

I expected to see this happen after creating documentation:

> Available on **crate feature a** only.

Instead, this happened:

> Available on **(crate features a or b or c) and crate feature a** only.

Occured when one module requires any of few features and then submodules requires one of those features - module should list **create features a or b or c** but submodule should return only **create feature a**.

### Meta

`rustc --version --verbose`:
```
rustc 1.67.0-nightly (a00f8ba7f 2022-11-15)
binary: rustc
commit-hash: a00f8ba7fcac1b27341679c51bf5a3271fa82df3
commit-date: 2022-11-15
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by fmease** [comment]
@rustbot label T-rustdoc F-doc_cfg

(note that you typo'ed `doc(cfg` as `cfg(doc(cfg` in your reproducer)

**Comment by ventaquil** [comment]
Oh, agree 👍 

My fault, original version was with `cfg_attr(docsrs, doc(...))` and copy-paste-delete in the middle of night needs more checks against bugs 🐛

**Comment by ventaquil** [comment]
Also when I have feature that enables all of `a`, `b` and `c` together, name it just `abc`.

```toml
[features]
a = []
b = []
c = []
abc = ["a", "b", "c"]
```

Then

```rust

#[doc(
    cfg(
        feature = "abc",
    )
)]
pub mod abc {
    #[doc(
        cfg(
            feature = "a",
        )
    )]
    pub mod a {
        // ...
    }
}
```

It will produce.

> Available on **crate feature abc and crate feature a** only.

Maybe it should return just it.

> Available on **crate feature a** only.

**Comment by jyn514** [comment]
Note this is a major expansion of scope, it requires a full SMT solver.

**Comment by est31** [comment]
It is correct that it's an expansion of scope, but it seems that SMT solvers aren't enough for this: I am currently still reading the literature but it seems it's not even in NP (like SMT) [but is Σ<sup>P</sup><sub>2</sub> - complete](https://doi.org/10.1016/j.jcss.2010.06.011) (a complexity class above NP).

The links I could find all relied on brute force which can be used less complex formulas but of course doesn't help with more complex ones:

* https://babbage.cs.qc.cuny.edu/courses/Minimize/
* https://research.swtch.com/boolean talks about a brute force approach that might be useful for small numbers of literals

For, SMT solvers there exist really performant Rust implementations, so we could just have used them as a library, but seems here we can't use them.

That being said, rustdoc is putting these notices all over the place so one should at least expect it to be able to do simple minimizations of the formula even if those don't arrive at the maximum possible level of minimization. Yes, it requires heavily algorithmic work. But the general skillset is present in Rust, and rustc is doing a lot of algorithmic stuff already in the type and borrow checkers (among others).

**Comment by jyn514** [comment]
I am very concerned about add large new features like that in rustdoc, especially without an RFC. The people working on the compiler are not the same people working on rustdoc (and historically it's been very hard to get the two to share knowledge).

**Comment by est31** [comment]
Fair points on the two groups of people being different. As for needing an RFC, a transformation of the form `(a or b or c) and a` -> `(b or c) and a` is trivial to implement, this shouldn't need an RFC imo. The Cargo.toml thing is harder as rustdoc would need the information about features implying each other passed to it via params. That might need an RFC indeed.

**Comment by ventaquil** [comment]
Referring to WolframAlpha `(a or b or c) and a` should return just `a`.

https://www.wolframalpha.com/input?i=simplify+%28a+or+b+or+c%29+and+a

**Comment by est31** [comment]
oh right good point. if a is false, then it's false, and if a is true then it's true. Anyways, it's easy to implement a few rules like these and apply them until a fix point is reached.

**Comment by Kixunil** [comment]
Wouldn't it be better to issue a warning in cases like this that some conditions do nothing? If someone wrote such condition it seems more likely it wasn't intended to contain contradictions or unused conditions. "Only for simple cases" still applies.

**Comment by Kixunil** [comment]
@est31 yes, that's a good example where warning is not suitable but simplification is. If one accidentally used `all` instead of `any` then a warning like this would be helpful:

> warning: `bar_only` is already gated on `feature= "bar"` in the module `hi`

**Comment by est31** [comment]
@Kixunil indeed that's one of the cases where the warning (i guess it would be a lint) would be useful. I think having both would be good, it's just not that warnings can provide a full replacement for doing it in rustdoc. Even in the instance where we issue a warning, rustdoc should still provide the simplification.

I think however that the two things, the warning and the rustdoc simplification, could share code, and if there is ever an implementation of the "feature foo implies bar" stuff, then they could both use the same CLI parameters. so the design/implementation burden on rustdoc would be minimized, which would also help with @jyn514 's concerns I think.

**Comment by est31** [comment]
Clippy's `nonminimal_bool` and `overly_complex_bool` lints use the [quine mc cluskey algorithm](https://en.wikipedia.org/wiki/Quine%E2%80%93McCluskey_algorithm) from [this](https://github.com/oli-obk/quine-mc_cluskey) repository.

**Comment by est31** [comment]
Looking closer at the `quine-mc_cluskey` crate and the [booleans.rs](https://github.com/rust-lang/rust-clippy/blob/7fe83edc11ccc255372be5a7bc0372d6011cc5df/clippy_lints/src/booleans.rs) file in clippy, it seems that there is some code around the usage of the `quine-mc_cluskey` crate. it's also limited to 8 variables only, as above that, performance starts to get really bad (it's an exponential algorithm due to the complexity mentioned above). There is also some filtering logic to check the output for solutions that contain a terminal more times than in the original version.

I've also filed a PR to make the crate a bit more up to date: https://github.com/oli-obk/quine-mc_cluskey/pull/5

**Comment by oli-obk** [comment]
We could definitely move some of the logic to the qmm crate and look into adding a generic API that avoids having to do the translation fully manually
